include/lo2s/line_info.hpp

```diff
@@ -32,33 +32,62 @@ namespace lo2s
 {
 struct LineInfo
 {
+private:
+    static const std::string unknown_file_str;
+    static const std::string unknown_binary_str;
+    static const std::string unmapped_function_str;
+    static const std::string unknown_function_str;
+
+    static constexpr unsigned int UNKNOWN_LINE = 0;
+
     // Note: If line is not known, we write 1 anyway so the rest is shown in vampir
-    LineInfo(const std::string& fi, const std::string& fu, unsigned int l, const std::string& d)
-    : file(fi), function(fu), line(l ? l : 1), dso(boost::filesystem::path(d).filename().string())
+    // phijor 2018-11-08: I think this workaround is not needed anymore? vampir
+    // shows source code locations with line == 0 just fine.
+    LineInfo(const std::string& file, const std::string& function, unsigned int line,
+             const std::string& dso)
+    : file(file), function(function), line((line != UNKNOWN_LINE) ? line : 1), dso(dso)
+    {
+    }
+
+public:
+    LineInfo() : LineInfo(unknown_file_str, unknown_function_str, UNKNOWN_LINE, unknown_binary_str)
+    {
+    }
+
+    static LineInfo for_function(const char* file, const char* function, unsigned int line,
+                                 const std::string& dso)
+    {
+        return LineInfo((file != nullptr) ? std::string{ file } : unknown_file_str,
+                        (function != nullptr) ? std::string{ function } : unknown_function_str,
+                        line, boost::filesystem::path(dso).filename().string());
+    }
+
+    static LineInfo for_unmapped_function()
```

Comment:
What is the difference to `for_unknown_function`?

Reply:
Can we use Meyer's singleton?

Reply:
> What is the difference to `for_unknown_function`?

It constructs with `unmapped_function_str` instead of `unknown_function_str`.

> Can we use Meyer's singleton?

Yes, that'd be nice. But right now I don't see how that's more efficient than the current pattern. I'll try to think of something.

Reply:
But we only have `unmapped_function/unknown_binary` and `unknown_function/${some binary}`?

Reply:
This is what I was thinking:

```suggestion
    static const LineInfo& for_unmapped_function() {
         static LineInfo li(unknown_file_str, unmapped_function_str, UNKNOWN_LINE, unknown_binary_str);
         return li;
    }
```

And then the question really was: Will this be copied from?

Reply:
> But we only have `unmapped_function/unknown_binary` and `unknown_function/${some binary}`?

So there's actually [this situation](https://github.com/tud-zih-energy/lo2s/pull/107/files#diff-1080ec2e46093f404fcf651ae4336d2aR594) where we have to "default construct" but I'm not quite sure what a sensible default is.

> And then the question really was: Will this be copied from?
As it stands, yes. `Trace::merge_ips` copies them.